[WIP][Testing] Mock Persistence Module in Utility Tests  - An Example

Olshansk opened the pull request:
## Description

This PR is a WIP and is meant to be a starting point with an example.

The goal is to enable the utility module to be developed while dependencies in the persistence module are not yet ready. This paradigm could extend to any other cross-module interaction.

The comments contain more detail, but in short:
1. A real instance of the persistence context is instantiated
2. A mocked instance of the persistence module instantiated and mocked
3. For some functions, we implement a passthrough from the mock -> real implementation
4. For other functions, we mock the behaviour; sometimes it's stateless and sometimes it's stateful

## Issue

Fixes #261

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other

## TODO: List of changes

- Change #1
- Change #2
- ...

## Testing
- [x] `go test -v -count=1 -v ./utility/... -run TestExperimental`
- [ ] `make develop_test`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

andrewnguyen22 commented on utility/test/module_test.go:
```diff
@@ -67,7 +66,19 @@ func NewTestingUtilityContext(t *testing.T, height int64) utility.UtilityContext
+	}
+}
+
+func newTestPersistenceModule(m *testing.M, databaseUrl string) modules.PersistenceModule {
```
Why do we still have the existing persistence module here? Is it safe to assume that this will be removed by this issue?

andrewnguyen22 commented on utility/test/experimental_test.go:
```diff
@@ -0,0 +1,207 @@
+	require.Equal(t, []byte(" 1  getHash  42  getHash "), appHash)
+}
+
+func newMockablePersistenceContextForAppHashTestWithMoreState(t *testing.T, height int64) modules.PersistenceRWContext {
```
Conversation around 'full mocking' vs 'hybrid' here https://github.com/pokt-network/pocket/pull/290#issuecomment-1270361284

Olshansk replied:
> like a 1:1 mapping of sorts?

Doesn't have to be a 1:1 mapping.

You can create helpers and then reuse them:

Imagine a `common_mocks_test.go` file
```go
    func sharedAddAccountAmount(address []byte, amount string) error {
       ... // implementation
    }
    
    // define other common functions
    
    sharedPersistenceContextMock := initializedPersistenceMock()
  
	sharedPersistenceContextMock.EXPECT().
		AddAccountAmount(gomock.Any(), gomock.Any()).
		DoAndReturn(sharedAddAccountAmount).
		AnyTimes()

    // expect other common functions
```

`some_test.go`

```go
    persistenceContextMock = copy(sharedPersistenceContextMock)
    
    // override some subset of mocks specific to this
```

`some_other_test.go`

```go
    persistenceContextMock = copy(sharedPersistenceContextMock)
    
    // override some other subset of mocks specific to this
```

Olshansk commented on utility/test/experimental_test.go:
```diff
@@ -0,0 +1,207 @@
+	"github.com/stretchr/testify/require"
+)
+
+// This test just tries to add amounts to an account and check the app hash via the real implementation
```
Yea, the way I've seen this done before is that you don't build it out all at once, but you add the functions/components you need over time.

Olshansk commented:
> This appears to not satisfy the encapsulation requirement

Yea, I see it as a step in the right direction.

> why don't we just Mock everything instead of using the persistence module?

I guess I have varying viewpoints here but don't feel strongly.

1. The persistence module has dozens of functions, and I was thinking that we could leverage existing business logic in the persistence module as an interim to unblock other work dependent on business logic that's not been implemented yet.

2. If we mock with placeholders, it makes sense, but if we mock with "real business logic", then we will pretty much need to maintain two complete versions of the persistence module. E.g., anytime we change something in the real persistence module, we'd need to change that here too. Could be the right path, but feels complex to me straight of the bat.

> It appears (atleast for the unit tests) we would be able to use the Expect statements to be able to test unlimited scenarios

Correct.

---

tl;dr I think what the "passthrough" approach does is unblock work that's dependant on functionality we don't have yet, without needing to pay the upfront cost of doing a complete second implementation of the persistence module. This may or may not be a good idea, but I think we will have visibility into it as we start mocking parts of it and get a clearer picture.

andrewnguyen22 commented:
@Olshansk seems like a decent amount of techdebt will be created if we go passthrough approach. I caution about the snowplow effect that @jessicadaugherty alluded to. 

The current open M3 issues for this sprint are not blocked by this, but the future ones will be. Since this is an XL task, I'd encourage us to go all the way with the Mock

Olshansk commented:
> Since this is an XL task, I'd encourage us to go all the way with the Mock

To make sure I understand, are you suggesting to maintain two "feature complete" versions of the persistence module, or just mocking everything in the persistence module interface with hardcoded values.

Will ping to sync offline as well.

andrewnguyen22 commented:
> just mocking everything in the persistence module interface with hardcoded values.

I'm suggesting to mock the entire persistence module in the Utility module to fully encapsulate rather than importing the persistence module and passing through

deblasis commented:
I have been thinking about this and I have a blurred idea that I wanted to share with you guys.

How about a function that constructs a mock and then multiple methods that in a fluent way allow us to configure expected behaviours?

Pseudo-code of a random scenario:

```go
persistenceContextMock:=newPersistenceContextMock(ctrl).
FromRuntimeMgr(runtimeMgr).
ShouldFailWhenGetAllFishermenIsCalledAtHeight(2, fmt.Errorf("some expected failure here").
AppHashShouldReturnAtHeight(2, []byte("somehash"))

persistenceMock:=newPersistenceModuleMock(ctrl).
WithPersistenceContextMock(persistenceContextMock).
SomeOtherBehaviouralMethod(...some_arguments)
```

And these can also be wrapped in some sort of util if we call them all the time and change only a bunch of parameters.

so in `Utility` you would call

```go
persistenceModuleMock := getHappyPathXXPersistenceModule()
```

It might be more complex than it sounds especially because probably there's gonna be some state to be handled inside the mock itself 🤔 but we could leverage the built-in controller to keep track of calls instead. I need to read more in depth how it all works.

Olshansk commented:
_Heads up: I added the typical Olshansky whitespaces in `utility/test/account_test.go` to help me understand the tests better._

@deblasis I believe the implementation I pushed is a starting point to what you described so PTAL.

@andrewnguyen22 I removed the code for my initial passthrough attempt and updated the unit tests in `utility/test/account_test.go` to use the new approach. 

Note how in places where we have dependencies on accounts, I do:

```go
	runTimeMgr := persistenceRuntimeMgr(t)
	acc := GetAllTestingAccounts(t, ctx)[0]
	ctx := NewTestUtilityContext(t, 0, withBaseAccountMock(t, runTimeMgr))
```

And in some places where we have dependencies on pools, I do:

```go
	runTimeMgr := persistenceRuntimeMgr(t)
	pool := GetAllTestingPools(t, ctx)[0]
	ctx := NewTestUtilityContext(t, 0, withBasePoolMock(t, runTimeMgr))
```

What the current implementation achieves:
1. Complete separation from the persistence module
2. Composable "building blocks" for a stateful persistence module mock that can be used in utility. E.g. you could include both pools AND accounts by running `NewTestUtilityContext(t, 0, withBasePoolMock(t, runTimeMgr), withBaseAccountMock(t, runTimeMgr))` if you needed to.
3. `withBasePoolMock` and `withBaseAccountMock` are stateful mocks, BUT we can explicitly EXPECT other functions with specific (i.e. not `gomock.Any()` values when we have more complex tests verifying specific flows

Please note that I do not believe this is "simple" or "low cognitive load" but fits the requirements of isolation by mocking the `shared` modules everywhere and sets up up to have other building blocks we can add in the future.

Let me know if jumping on a call would help instead.

Olshansk commented:
@andrewnguyen22 Wanted to bump this for you to provide some feedback if this is what you had in mind.

andrewnguyen22 reviewed (changes requested):
DUP COMMENT SEE BELOW

Olshansk commented:
Closing out this PR for now, but we can use it as a reference for future work on #261.
